feat: course outline - units list

navinkarkera opened the pull request:
**Description:**
This PR adds units to course outline with duplicate, delete, publish and adding new unit functionality.

**Screenshot:**
![image](https://github.com/openedx/frontend-app-course-authoring/assets/10894099/cfb84b5d-11a2-4cdd-a775-518e15737810)


**Test instructions:**

* Setup this repo as described in the readme.
* Checkout branch from https://github.com/openedx/edx-platform/pull/33943 and run `make cms-up` in devstack.
* Open any course outline by using this url format: http://localhost:2001/course/course-v1:edX+DemoX+Demo_Course/
* Test units delete, publish, new and publish functionality by expanding any subsection.

Cup0fCoffee commented on src/course-outline/card-header/TitleButton.jsx:
```diff
@@ -0,0 +1,76 @@
+        className="item-card-header__expanded-btn"
+        onClick={onTitleClick}
+      >
+        <Truncate lines={1} className={`${namePrefix}-card-title mb-0`}>{title}</Truncate>
+        {badgeTitle && (
+          <div className="item-card-header__badge-status" data-testid={`${namePrefix}-card-header__badge-status`}>
+            {badgeIcon && (
+              <Icon
+                src={badgeIcon}
+                size="sm"
+                className={classNames({ 'text-success-500': status === ITEM_BADGE_STATUS.live })}
+              />
+            )}
+            <span className="small">{badgeTitle}</span>
+          </div>
+        )}
```
@navinkarkera The number of lines of code seems little, but there are many details in them. I feel like it is enough to extract it into a base component `Title` (or `BaseTitle`), to avoid duplication and having to mirror the changes in two places.

Cup0fCoffee reviewed (approved):
:+1: I made some suggestions, but otherwise LGTM

- [x] I tested this: created new sections, subsections, units and used the action drop down to duplicate, publish and delete the units
- [x] I read through the code

navinkarkera commented on src/course-outline/unit-card/UnitCard.scss:
```diff
@@ -0,0 +1,27 @@
+    display: none;
+  }
+
+  .unit-card__content {
```
It is used in `src/course-outline/card-header/TitleButton.jsx` and `src/course-outline/card-header/TitleLink.jsx` as `${namePrefix}-card-title`. Added a comment above the selector.

navinkarkera commented on src/course-outline/CourseOutline.test.jsx:
```diff
@@ -271,177 +292,272 @@ describe('<CourseOutline />', () => {
+          },
+        });
+
+      const editButton = await within(element).findByTestId(`${elementName}-edit-button`);
+      fireEvent.click(editButton);
+      const editField = await within(element).findByTestId(`${elementName}-edit-field`);
+      fireEvent.change(editField, { target: { value: newName } });
+      await act(async () => fireEvent.blur(editField));
+      expect(
+        axiosMock.history.post[axiosMock.history.post.length - 1].data,
+        `Failed for ${elementName}!`,
+      ).toBe(JSON.stringify({
         metadata: {
-          display_name: newDisplayName,
+          display_name: newName,
         },
-      }))
-      .reply(200, { dummy: 'value' });
-    axiosMock
-      .onGet(getXBlockApiUrl(section.id))
-      .reply(200, {
-        ...section,
-        display_name: newDisplayName,
-      });
-
-    const [sectionElement] = await findAllByTestId('section-card');
-    const editButton = await within(sectionElement).findByTestId('section-edit-button');
-    fireEvent.click(editButton);
-    const editField = await within(sectionElement).findByTestId('section-edit-field');
-    fireEvent.change(editField, { target: { value: newDisplayName } });
-    await act(async () => {
-      fireEvent.blur(editField);
-    });
-
-    expect(await findByText(newDisplayName)).toBeInTheDocument();
-  });
+      }));
+      const results = await within(element).findAllByText(newName);
+      expect(results.length, `Failed for ${elementName}!`).toBeGreaterThan(0);
+    };
```
@Cup0fCoffee I did try it, but it requires us to repeat the queries for each type and it increases test time due to re-rendering of the whole outline for each type. I am also not sure whether this is the best way forward but to improve logging I included [jest-expect-message](https://www.npmjs.com/package/jest-expect-message).

navinkarkera commented:
@Cup0fCoffee Thanks for the nice suggestions! Applied all of them.

navinkarkera commented:
@KristinAoki This is ready for you.

KristinAoki reviewed (changes requested):
Additionally to these changes, it would good to add a loading spinner to the page while the sections are loading.

navinkarkera commented on src/course-outline/hooks.jsx:
```diff
@@ -68,6 +74,26 @@ const useCourseOutline = ({ courseId }) => {
     dispatch(addNewSubsectionQuery(sectionId));
   };
 
+  const getUnitUrl = (locator) => {
+    if (process.env.ENABLE_UNIT_PAGE === 'true') {
+      return `/course/container/${locator}`;
+    }
+    return `${getConfig().STUDIO_BASE_URL}/container/${locator}`;
```
@KristinAoki If we always go through external link, we will loose the speed of page change via react-router and refresh the whole page even if units page is part of same MFE.

navinkarkera commented on src/course-outline/subsection-card/SubsectionCard.jsx:
```diff
@@ -97,23 +115,23 @@ const SubsectionCard = ({
+            onClick={handleNewButtonClick}
           >
             {intl.formatMessage(messages.newUnitButton)}
           </Button>
```
Updated style of all new-* buttons to match figma design.

navinkarkera commented:
> Additionally to these changes, it would good to add a loading spinner to the page while the sections are loading.

@KristinAoki Done. Addressed all your comments except for updating the unit page link as I have a small doubt there.

KristinAoki commented on src/course-outline/CourseOutline.jsx:
```diff
@@ -235,7 +235,7 @@ const CourseOutline = ({ courseId }) => {
                           <Button
                             data-testid="new-section-button"
-                            className="mt-4"
+                            className="mt-4 bg-light-100 border-gray-200"
```
Remove the background and border color. The hover state when these are changed cause an accessibility contrast error. The default styling for the `outline-primary` variant matches the design in figma. This also applies to the new subsection and new unit buttons.

navinkarkera replied:
@KristinAoki The designs show white background for new section and new unit button but the default styling for `outline-primary` variant has transparent background so it becomes grayish.

KristinAoki replied:
That is ok, because it is the expected behavior for the open edx button styles. The mocks are built using the edx styles so not everything is going to match if you use the default open edx styles.

navinkarkera replied:
@KristinAoki Got it. Updated.

navinkarkera commented:
@KristinAoki Thanks! Squashed all commits, ready to be merged.

openedx-webhooks commented:
@navinkarkera 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-course-authoring&entry.867055334=https://github.com/openedx/frontend-app-course-authoring/pull/765&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2023-12-22+11:00&entry.1917517419=2024-01-10+14:38&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
